software/comms_module/SmartBandage/Application/fsm.c

```diff
@@ -1,139 +1,229 @@
+    //no connection between the communications module and the bandage module detected
+    //set the current state to temp error and wait for check timer to expire to see if initialized (?)
+	return SB_switchState(S_ERROR_TEMP);
+}
+SB_State SB_errorEvent(){
+    //a more permanent error has occurred, do nothing or else relay problem via bluetooth
+
+	//permanent errors
+	switch (systemState.currentError) {
+	case OSResourceInitializationError:
+	case OutOfMemory:
+		return SB_switchState(S_ERROR_PERM);
+
+	// All other errors implied temporary
+	default:
+		return SB_switchState(S_ERROR_TEMP);
+	}
+}
+SB_State SB_dataChanged(void){
+    //data changed prior to the ble_timer expiring
+	return SB_switchState(S_TRANSMIT);
+}
+
+//not actually an event, just something that happens when there isn't anything else to do
+SB_State SB_returnToSleep(void){
+	return SB_switchState(S_SLEEP);
+}
+
+SB_State SB_handleEvent(SB_Event event) {
+	// State or event invalid
+	if (systemState.currentState >= SB_NUM_STATES || event >= SB_NUM_EVENTS) {
+		SB_setError(InvalidParameter);
+		return SB_errorEvent();
+	}
+
+	// No action if the pointer is null
+	if (SB_stateTable[systemState.currentState][event] == NULL) {
+		return systemState.currentState;
+	}
+
+	return SB_stateTable[systemState.currentState][event]();
+}
+
+inline void SB_setError(SB_Error error) {
+	systemState.lastError = systemState.currentError;
+	systemState.currentError = error;
+}
+
+inline SB_State SB_currentState() {
+	return systemState.currentState;
+}
+
+
+// Called from within periheral functions to register that the peripheral will need to be revisited when the state changes
+void SB_registerStateTransitionCallback(void function(void), SB_State_Transition peripheral) {
+	//create the callback
+	SB_CallbackFunc *newCallback;
+	newCallback = (SB_CallbackFunc*)malloc( sizeof(SB_CallbackFunc));
+	newCallback->transition = peripheral;
+	newCallback->function = function;
+	newCallback->next = 0; //start with the new callback not pointing to anything
+	//add the callback to the transition table
+	SB_addCallback(newCallback);
+}
+
+/*
+void SB_registerEvent(SB_EventHandler eventHandler, SB_State state) {
+	//passes function pointer to be called in the future
+	(*eventHandler)();
+}
+*/
+
+
+// This will be put within the switchstate cases to make sure the peripherals are dealt with correctly
+// This will be put within the switchstate cases to make sure the peripherals are dealt with correctly
+void SB_callCallback( SB_State_Transition state) {
+	//check if the linked list in empty, if it is then don't do anything, if it isn't, then iterate through list and call functions
+    SB_CallbackFunc *current;
+	if(callbackTable == NULL) {
+        return;
+	}
+	current = callbackTable->callbacks ;
+	// disable context switching for other tasks in this section
+	taskKey = Task_disable();
+	while(current->next != NULL) {
+		if(state == current->transition){
+			current->function();
```

Comment:
Always a good idea to check pointers before you execute them - even though nothing nefarious should happen with the current code.